std/math/big/int.zig

```diff
@@ -0,0 +1,2023 @@
+    }
+
+    pub fn ensureCapacity(self: *Int, capacity: usize) !void {
+        if (capacity <= self.limbs.len) {
+            return;
+        }
+
+        self.limbs = try self.allocator.realloc(Limb, self.limbs, capacity);
+    }
+
+    pub fn deinit(self: *const Int) void {
+        self.allocator.free(self.limbs);
+    }
+
+    pub fn clone(other: *const Int) !Int {
+        return Int{
+            .allocator = other.allocator,
+            .positive = other.positive,
+            .limbs = block: {
+                var limbs = try other.allocator.alloc(Limb, other.len);
+                mem.copy(Limb, limbs[0..], other.limbs[0..other.len]);
+                break :block limbs;
+            },
+            .len = other.len,
+        };
+    }
+
+    pub fn copy(self: *Int, other: *const Int) !void {
+        if (self == other) {
+            return;
+        }
+
+        self.positive = other.positive;
+        try self.ensureCapacity(other.len);
+        mem.copy(Limb, self.limbs[0..], other.limbs[0..other.len]);
+        self.len = other.len;
+    }
+
+    pub fn swap(self: *Int, other: *Int) void {
+        mem.swap(Int, self, other);
+    }
+
+    pub fn dump(self: *const Int) void {
+        for (self.limbs) |limb| {
+            debug.warn("{x} ", limb);
+        }
+        debug.warn("\n");
+    }
+
+    pub fn negate(r: *Int) void {
+        r.positive = !r.positive;
+    }
+
+    pub fn abs(r: *Int) void {
+        r.positive = true;
+    }
+
+    pub fn isOdd(r: *const Int) bool {
+        return r.limbs[0] & 1 != 0;
+    }
+
+    pub fn isEven(r: *const Int) bool {
+        return !r.isOdd();
+    }
+
+    fn bitcount(self: *const Int) usize {
+        const u_bit_count = (self.len - 1) * Limb.bit_count + (Limb.bit_count - @clz(self.limbs[self.len - 1]));
+        return usize(!self.positive) + u_bit_count;
+    }
+
+    pub fn sizeInBase(self: *const Int, base: usize) usize {
+        return (self.bitcount() / math.log2(base)) + 1;
+    }
+
+    pub fn set(self: *Int, value: var) Allocator.Error!void {
+        const T = @typeOf(value);
+
+        switch (@typeInfo(T)) {
+            TypeId.Int => |info| {
+                const UT = if (T.is_signed) @IntType(false, T.bit_count - 1) else T;
+
+                try self.ensureCapacity(@sizeOf(UT) / @sizeOf(Limb));
+                self.positive = value >= 0;
+                self.len = 0;
+
+                var w_value: UT = if (value < 0) UT(-value) else UT(value);
+
+                if (info.bits <= Limb.bit_count) {
+                    self.limbs[0] = Limb(w_value);
+                    self.len = 1;
+                } else {
+                    var i: usize = 0;
+                    while (w_value != 0) : (i += 1) {
+                        self.limbs[i] = @truncate(Limb, w_value);
+                        self.len += 1;
+
+                        // TODO: shift == 64 at compile-time fails. Fails on u128 limbs.
+                        w_value >>= Limb.bit_count / 2;
```

Comment:
There are a few deficiencies in the stage1 compiler's big integer library. I'll make a pass through at some point and fix up the edge cases as these issues typically result in silent comptime errors which is the worst kind.